torch/lib/THD/base/data_channels/DataChannelNccl.cpp

```diff
@@ -0,0 +1,731 @@
+#include "../Cuda.hpp"
+#include "../../../csrc/utils/auto_gpu.h"
+#include "DataChannelNccl.hpp"
+#include "DataChannelUtils.hpp"
+
+#include <cuda.h>
+#include <THC/THC.h>
+
+#include <unistd.h>
+
+#include <cstdint>
+#include <stdexcept>
+#include <unordered_set>
+#include <sstream>
+
+namespace thd {
+
+namespace {
+
+
+std::unordered_map<THDReduceOp, ncclRedOp_t> ncclOp = {
+  {THDReduceOp::THDReduceMIN, ncclMin},
+  {THDReduceOp::THDReduceMAX, ncclMax},
+  {THDReduceOp::THDReduceSUM, ncclSum},
+  {THDReduceOp::THDReducePRODUCT, ncclProd},
+};
+
+
+std::unordered_map<at::ScalarType, ncclDataType_t> ncclDatatype = {
+  {at::kChar, ncclInt8},
+  {at::kByte, ncclUint8},
+  {at::kFloat, ncclFloat},
+  {at::kDouble, ncclDouble},
+  {at::kInt, ncclInt32},
+  {at::kLong, ncclInt64},
+};
+
+
+// Helper function that gets the data type and issues error if not supported
+static ncclDataType_t _getNcclDataType(at::ScalarType type) {
+  if (ncclDatatype.find(type) == ncclDatatype.end()) {
+    throw std::runtime_error("Unsupported data type for NCCL");
+  }
+  return ncclDatatype[type];
+}
+
+
+// Helper function that gets the device list to determine the CUDA devices
+std::vector<int> getDevicesList(const std::string& deviceSeq) {
+
+  std::stringstream ss(deviceSeq);
+  std::string device;
+  std::vector<int> devices;
+  while (std::getline(ss, device, ',')) {
+    devices.push_back(stoi(device));
+  }
+  return devices;
+}
+
+} // namespace
+
+
+// RequestNccl
+DataChannelNccl::RequestNccl::RequestNccl(QueueWorker::Request&& request)
+ : _request(std::move(request)) {
+}
+
+
+DataChannelNccl::RequestNccl::~RequestNccl() {}
+
+
+bool DataChannelNccl::RequestNccl::isCompleted() {
+  return _request.isCompleted();
+}
+
+void DataChannelNccl::RequestNccl::wait() {
+  _request.wait();
+}
+
+// End of RequestNccl
+
+// DataChannelNccl
+DataChannelNccl::DataChannelNccl(InitMethod::Config config, int timeout)
+  : _rank(config.rank)
+  , _numProcesses(config.world_size)
+  , _timeout(timeout)
+  , _masterListeningSocket(-1)
+{
+  if (_rank == 0) {
+    _masterListeningSocket = config.master.listen_socket;
+  } else {
+    _masterAddr = config.worker.master_addr;
+    _masterPort = config.worker.master_port;
+  }
+}
+
+
+// Use the socket to broadcast NCCL ID
+void DataChannelNccl::broadcastUniqueNcclId(ncclUniqueId* srcNcclId,
+                                            ncclUniqueId* dstNcclId) {
```

Comment:
But is it even used? It seems that only the first process broadcasts it, all other processes just ignore that argument.

Reply:
I just realized that you're always passing the same address twice, because it's only used for one of those purposes. Why not take in a single pointer? It would also let you call `ncclGetUniqueId` once inside this method only in first process (we discard all other values).

Reply:
Taking a single pointer is fine too.